### PR TITLE
#167730740 Build Email Verification feature on user sign up

### PR DESCRIPTION
**Pull request by zanio**
#### What does this PR do?
It Build Email Verification feature on user sign up and sends email for verification to user on successful sign up
### Description of Task to be completed?
N/A
#### How should this be manually tested?

1. ```git clone -b ft-email-verification-167730740  https://github.com/andela/firestar-backend.git .```  @[ft-email-verification-167730740](https://github.com/andela/firestar-backend/tree/ft-email-verification-167730740)
2. run `npm install` and create  a `.env file` and provide the following information to it 
```
EMAIL = 
LAYER=
PASSWORD=
SECRET_KEY_EMAIL_VERIFY_TOKEN =
PORT = 
```
 
4. start the server with `npm run start:dev`

5. Now use postman and go to `localhost:PORT/api/v1/users/email/test` using a `POST` request with a body of 
```json
body:{ 
email: 'youremail',
firstName: 'Yourfirstname'
lastName: 'Yourlastname'
}
```
6. If successful you would see a response similar to the response of the screenshot on postman
7. Visit the email you provided and see the newly delivered email which contains your name and link to verify your account.
8. copy to clipboard the `id key` token from the url sent to the provided mail and use it on postman on a `get` request on `http://localhost:PORT/api/v1/users/email/verify?id=token` to verify the token sent

#### Any background context you want to provide?
Please note that a token expires after 24hrs
#### What are the relevant pivotal tracker stories?
[#167730740](https://www.pivotaltracker.com/story/show/167730740)

#### Screenshots or gifs (if appropriate)
**Test coverage information**

image

* 
Screenshot (29)

image

#### Questions:
* N/A

### CODE REVIEW

**Comment by wahabtobibello** [comment]
@rovilay Shouldn't automated tests be written for this?

**Comment by zanio** [comment]
> @rovilay Shouldn't automated tests be written for this?

@rovilay what do you have to say

**Comment by OlawaleJoseph** [comment]
Well done Bro. I tried testing the functionality on my machine but got an error
Screenshot (16)

Also, I noticed that:
 -  Travis is failing
 - Unresolved Hound Issues

**Review by ola357 — COMMENTED** [review]
Excellent work, 
1. Travis is failing, please fix the tests. Fix the hound issues too
2. I don't think you should return a 500 error in case of a fail

**Comment by zanio** [comment]
> Well done Bro. I tried testing the functionality on my machine but got an error
> Screenshot (16)
> 
> Also, I noticed that:
> 
> * Travis is failing
> * Unresolved Hound Issues

@ola357 i observed this was runing on port 9900? Mine is runing on port 3000. also please make sure the request body looks like the screenshot below.



Screenshot (30)

**Review by wahabtobibello — APPROVED** [review]

**Review by dewaleolaoye — APPROVED** [review]
Well Done!
Removing `.env` file from `gitignore` exposes so many information like `email`, `password`, and `token`.
I suggest you include these details in the PR thread you posted on slack

File: .gitignore [hunk 1]
```diff
@@ -72,7 +71,7 @@ typings/
 .yarn-integrity
 
 # dotenv environment variables file
-.env
```
Comment (rovilay) [inline .gitignore]: Why remove `.env` from `.gitignore`?

File: .gitignore [hunk 2]
```diff
@@ -72,7 +71,7 @@ typings/
...
+
 .env.test
```
Comment (rovilay) [inline .gitignore]: is there a reason for adding this `.env.test`?

File: package.json [hunk 3]
```diff
@@ -59,13 +60,16 @@
...
     "coveralls": "^3.0.6",
     "eslint": "^6.1.0",
     "eslint-config-airbnb-base": "^14.0.0",
     "eslint-plugin-import": "^2.18.2",
     "mocha": "^6.2.0",
     "mocha-lcov-reporter": "^1.3.0",
     "nodemon": "^1.19.1",
-    "nyc": "^14.1.1"
+    "nyc": "^14.1.1",
+    "sinon": "^7.4.1",
```
Comment (rovilay) [inline package.json:72]: Why `sinon`?

File: src/__tests__/emailverification.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    it('it should not assign token, or fail', async () => {
+      const id = 'some_secret_identification_number_or_string';
+      const token = await emailVerifyToken(id);
+      expect(token).to.equal(token);
+    }).timeout(0);
+  });
+
+  describe('EMAIL TOKEN CONFIRMATION ROUTE', () => {
+    it('should have a status of 200 when valid token is sent as query string', async () => {
+      const id = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InNvbWVfZW5jb2RlZF9pZGVudGlpdHkiLCJpYXQiOjE1NjY4MTE1OTcsImV4cCI6MTU2Njg5Nzk5N30.oCXhqT4Ri-k7RIqlgIHVFvGGwmxIzBwyzqbXIG0JwyE';
```
Comment (rovilay) [inline src/__tests__/emailverification.js]: can you define the `id` value in a `__mocks__` folder and import it here since you're using it in multiple places.

File: src/controllers/user.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+
+/* eslint-disable import/prefer-default-export */
+import jwt from 'jsonwebtoken';
```
Comment (rovilay) [inline src/controllers/user.js]: Please make your controllers class-based. Reach out to @conquext.

File: src/routes/api/users.js [hunk 6]
```diff
@@ -1,82 +1,11 @@
...
-  User.findById(req.payload.id)
-    .then((user) => {
-      if (!user) {
-        return res.sendStatus(401);
-      }
-      return res.json({ user: user.toAuthJSON() });
-    })
-    .catch(next);
-});
+router.post('/email/test', SendVerificationEmail, signUp);
```
Comment (rovilay) [inline src/routes/api/users.js]: please version your routes i.e `/api/v1/email`

Comment (rovilay) [inline src/routes/api/users.js]: Also why `email/test`? what is the purpose?

File: src/services/mail/Mail.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
+/* eslint-disable import/prefer-default-export, no-console, require-jsdoc */
+
+import nodemailer from 'nodemailer';
+import dotenv from 'dotenv';
+
+dotenv.config();
+
+
```
Comment (rovilay) [inline src/services/mail/Mail.js]: Multiple spaces

File: src/services/mail/Mail.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
+/* eslint-disable import/prefer-default-export, no-console, require-jsdoc */
```
Comment (rovilay) [inline src/services/mail/Mail.js]: Remove comments and fix eslint errors.

File: src/services/mail/Mail.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+class Mail {
+  constructor(emailDetails, html) {
+    this.subject = emailDetails.Subject;
+    this.recipient = emailDetails.Recipient;
+    this.content = html;
+    this.Recipients = this.formatRecipients();
+  }
+
+  // When sending to multiple users;
```
Comment (rovilay) [inline src/services/mail/Mail.js]: JsDoc this.

File: src/utils/index.js [hunk 10]
```diff
@@ -1,2 +1,14 @@
...
+const emailVerifyToken = (id) => new Promise((resolve) => {
+  const token = jwt.sign({ id }, process.env.SECRET_KEY_EMAIL_VERIFY_TOKEN, { expiresIn: '24h' });
+  if (token) resolve(token);
+});
+
 const arrayTest = (array, index) => array.indexOf(index);
-export default arrayTest;
+
+
+export { emailVerifyToken, arrayTest };
```
Comment (rovilay) [inline src/utils/index.js]: Please move this ticket back to `peer review` while Implementing feedback.

File: .gitignore (hunk 1 above)
Comment (zanio) [reply .gitignore]: I don't have access to the Build environment to set the environment variable upon deployment. When i `gitignore` the `.env` file it only ran locally on my pc, but failed upon deploying.

File: .gitignore (hunk 2 above)
Comment (zanio) [reply .gitignore]: Nope. It was not added by me.

File: package.json (hunk 3 above)
Comment (zanio) [reply package.json:72]: It is used to check for what a function returns when it is called. Kind of faking a function.

File: src/routes/api/users.js (hunk 6 above)
Comment (zanio) [reply src/routes/api/users.js]: Okay, done

File: .gitignore (hunk 2 above)
Comment (zanio) [reply .gitignore]: I would remove it

File: .gitignore (hunk 1 above)
Comment (rovilay) [reply .gitignore]: @zanio Yes, that is how it's supposed to be. You're supposed to set the `.env` variables in your production/deployment environment.

File: .gitignore (hunk 2 above)
Comment (rovilay) [reply .gitignore]: don't remove it if it wasn't added by you.

File: package.json (hunk 3 above)
Comment (rovilay) [reply package.json:72]: OK.

File: src/__tests__/emailverification.js (hunk 4 above)
Comment (zanio) [reply src/__tests__/emailverification.js]: Alright, that would be done

File: src/routes/api/users.js (hunk 6 above)
Comment (zanio) [reply src/routes/api/users.js]: Since i am implementing just the email verification feature, i felt having just a separate route would be fine. so when the signup feature is ready the `/email/test` middleware ( `SendVerificationEmail` ) would be passed as one of the middleware to signup

File: src/services/mail/Mail.js (hunk 7 above)
Comment (zanio) [reply src/services/mail/Mail.js]: Done

File: .gitignore (hunk 1 above)
Comment (zanio) [reply .gitignore]: I would therefore need access on firestar-backend heroku to add this variable

File: .gitignore (hunk 2 above)
Comment (zanio) [reply .gitignore]: Alright

File: src/controllers/user.js (hunk 5 above)
Comment (zanio) [reply src/controllers/user.js]: Alright, That would be done

File: src/utils/index.js (hunk 10 above)
Comment (zanio) [reply src/utils/index.js]: Okay

File: .env.example [hunk 11]
```diff
@@ -1,7 +0,0 @@
-DB_USERNAME=postgres
```
Comment (rovilay) [inline .env.example]: Why did you delete this file?

Comment (zanio) [reply .env.example]: I felt it's not needed since i already gitignore the `.env file` already which my feature is using

File: .gitignore (hunk 1 above)
Comment (rovilay) [reply .gitignore]: ok

File: .env.example (hunk 11 above)
Comment (rovilay) [reply .env.example]: We still need it. It's what shows anybody that will work on the project the `.env` needed to get the app running.

File: .env.example [hunk 12]
```diff
@@ -2,6 +2,5 @@ DB_USERNAME=postgres
 DB_PASSWORD=db_password
 DB_NAME=barefoot_nomad
 DB_HOST=127.0.0.1
-PORT=
```
Comment (rovilay) [inline .env.example:5]: Why did you remove the `PORT`?

File: .env.example (hunk 11 above)
Comment (zanio) [reply .env.example]: Oh, Alright. I would do likewise

File: .env.example (hunk 12 above)
Comment (zanio) [reply .env.example:5]: Well now i understand the use of `env.example`.

**Comment by rovilay** [comment]
@zanio Good work so far. For your line 5 and 8 of your PR description please change the `3000` to `<PORT>`, that way users can know that the value is not fixed but depends on your `PORT` variable.

File: src/middlewares/mail.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
+/* eslint-disable import/prefer-default-export */
+import { verifyEmail } from '../services/mail/template/verifyEmail';
+import { emailVerifyToken } from '../utils/index';
+import Mail from '../services/mail/Mail';
+
+export const SendVerificationEmail = async (req, res, next) => {
+  let { email, firstName, lastName } = req.body;
+  email = email ? email.trim() : '';
+  firstName = firstName.trim();
```
Comment (rovilay) [inline src/middlewares/mail.js]: @zanio When I tested your `email/test` endpoint, I noticed you didn't validate the `req.body` and you didn't handle errors too. I noticed errors on my console. Please handle errors for cases when the req.body isn't present or the req payload is incomplete or invalid.
Screenshot 2019-08-28 at 10 11 34 AM

File: src/middlewares/mail.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+  lastName = lastName.trim();
+  const id = 'some_encoded_identiity';
+  const token = await emailVerifyToken(id);
+  const emaildDetails = {
+    Subject: 'Email Verification',
+    Recipient: email,
+  };
+  const link = `http://localhost:3000/api/v1/auth/verify?id=${token}`;
```
Comment (rovilay) [inline src/middlewares/mail.js]: this link is wrong. it shows you using `/auth/verify` but the app uses `/email/verify`. Also, I believe the endpoint should be `/users/email/test` and `/users/email/verify` since its more user-focused and it's in the users' route folder. Also, the `PORT` should be dynamic because a user might use a different `PORT` i.e use the `.env PORT`.
@dewaleolaoye  @wahaaabello how did you miss this?

File: src/middlewares/mail.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+  const data = {
+    email,
+    firstName,
+    lastName,
+    link,
+  };
+
+  try {
+    const send = new Mail(emaildDetails, verifyEmail(data));
```
Comment (rovilay) [inline src/middlewares/mail.js]: @zanio the `verifyEmail` function should be better named to denote what it's actually doing which is to generate a `verify email template` hence rename to `verifyEmailTemplate`

**Comment by rovilay** [comment]
@zanio Please update your tests to cover for the feedback.

File: src/middlewares/mail.js (hunk 14 above)
Comment (zanio) [reply src/middlewares/mail.js]: Fixed

File: src/middlewares/mail.js (hunk 15 above)
Comment (zanio) [reply src/middlewares/mail.js]: Fixed

**Comment by zanio** [comment]
> @zanio Please update your tests to cover for the feedback.

@rovilay Its now updated 9 successful test made

File: src/index.js [hunk 16]
```diff
@@ -19,6 +19,7 @@ const app = express();
 
 // swagger config middlewares
 app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));
+app.enable('trust proxy');
```
Comment (rovilay) [inline src/index.js:22]: Why this?

File: src/middlewares/mail.js [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * @param {req} req contains the express object.
+ * @param {res} res contains the express response.
+ * @param {next} next calls the next middleware.
+ * @returns {next-middleware } returns the next middleware or throws an error.
+ * @returns {error} if any error occurs it throws an error.
+ */
+export const SendVerificationEmail = async (req, res, next) => {
+  let { email, firstName, lastName } = req.body;
```
Comment (rovilay) [inline src/middlewares/mail.js:16]: @zanio Good work so far. Although, you handled the errors but it not in an efficient way. Before sending the `Mail`,  you should `verify` the required fields which are `email, firstName and lastName`. But from your implementation, you just append an empty string to the parameters `''` then wait for the `send.main()` to fail before returning a message. That is not efficient please rectify.

File: src/middlewares/mail.js [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+  try {
+    const send = new Mail(emaildDetails, verifyEmailTemplate(data));
+    const response = await send.main();
+    if (!response) {
+      return res.status(400).json({ status: 400, error: 'Network error occured, please check your network' });
+    }
+
+    if (response.message === 'No recipients defined') {
+      return res.status(409).json({ status: 409, error: response.message });
```
Comment (rovilay) [inline src/middlewares/mail.js:51]: Why is the response status `409`?

File: src/index.js (hunk 16 above)
Comment (zanio) [reply src/index.js:22]: `app.enable('trust proxy');` enables our application to use `https` . By default `req.protocols = http` which is not a protected protocol. So when that is used it enables our server to use secured protocol. Why this was applied was in the generation of verification link. When we are on our local environment it evaluates to false and true when in production. So i used it in the construction of my verification link. for local enviroment we have something like `http://localhost...` and for production we have `https://...` To see this in application make a post request from `https://firestar-backend-staging-pr-24.herokuapp.com/api/v1/users/email/test` and visit your mail inbox you would observe the link on that mail received has a secured protocol  `https`

File: src/middlewares/mail.js (hunk 17 above)
Comment (zanio) [reply src/middlewares/mail.js:16]: Done

**Review by ola357 — APPROVED** [review]
welldone, tested it okay

File: .env.example [hunk 19]
```diff
@@ -2,6 +2,9 @@ DB_USERNAME=postgres
...
-
 DB_TEST_NAME=barefoot_test
+PORT = 
+EMAIL = 
+LAYER =
```
Comment (rovilay) [inline .env.example]: what is the use of `LAYER` env?

File: .gitignore [hunk 20]
```diff
@@ -24,6 +23,7 @@ pids
 *.pid.lock
 package-lock.json
 
+
```
Comment (rovilay) [inline .gitignore]: multiple space.

File: package.json [hunk 21]
```diff
@@ -4,7 +4,7 @@
   "description": "A Social platform for the creative at heart",
   "main": "index.js",
   "scripts": {
-    "test": "cross-env NODE_ENV=development nyc mocha --require @babel/register ./src/__tests__/*.js",
+    "test": "cross-env NODE_ENV=development nyc mocha --require @babel/register --require @babel/polyfill -r dotenv/config ./src/__tests__/*.js --exit",
```
Comment (rovilay) [inline package.json:7]: what does the `dotenv` config do?

File: src/__tests__/emailverification.js [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+      expect(response.body).to.be.a('object');
+    }).timeout(0);
+
+    it('should have a status of 403 when BODY is not present', async () => {
+      const response = await request.post('/api/v1/users/email/test');
+      expect(response.body.status).to.equal(403);
+      expect(response.body).to.be.a('object');
+    }).timeout(0);
+
+    it('should have a status of 403 and a messsage of "please fill in the required inputs"  when some body is not present', async () => {
```
Comment (rovilay) [inline src/__tests__/emailverification.js]: the message `please fill in the required inputs` is too vague. you should try and be more specific i.e `email, firstname is/are required`. This will help users solve the issue faster.

File: src/__tests__/emailverification.js [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+      };
+      const response = await request.post('/api/v1/users/email/test').send(body);
+      expect(response.body.error).to.equal('Please fill in the required inputs');
+      expect(response.body.status).to.equal(403);
+      expect(response.body).to.be.a('object');
+    }).timeout(0);
+
+    it('should have a status of 403 and a messsage of "Please provide a valid email"  when some body is not present', async () => {
+      const body = {
+        email: 'akp.aniyahoo.com',
```
Comment (rovilay) [inline src/__tests__/emailverification.js:66]: This is good. 👍

File: src/__tests__/emailverification.js [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+    it('should have a status of 200 when valid token is sent as query string', async () => {
+      const id = tokenEmail;
+      const response = await request.get(`/api/v1/users/email/verify?id=${id}`);
+      expect(response.body.status).to.equal(200);
+      expect(response.body).to.be.a('object');
+    }).timeout(0);
+  });
+
+  describe('EMAIL TOKEN CONFIRMATION ROUTE', () => {
+    it('should have a status of 404 when invalid token is sent as query string', async () => {
```
Comment (rovilay) [inline src/__tests__/emailverification.js]: Are you testing for `404` or `400`?

File: .env.example (hunk 19 above)
Comment (zanio) [reply .env.example]: Nodemailer package require you provide the `LAYER` which is the sender mailling service provider. It can be `e.g include Yahoo, gmail` and your private account mail provider as well.we intend

File: src/middlewares/mail.js (hunk 18 above)
Comment (rovilay) [reply src/middlewares/mail.js:51]: @zanio 👆

File: src/validation/emailValidation.js [hunk 25]
```diff
@@ -0,0 +1,9 @@
+/* eslint-disable import/prefer-default-export */
+import { regexForEmail } from '../__mocks__/emailVerification';
```
Comment (rovilay) [inline src/validation/emailValidation.js]: abstracting `regexForEmail` is a good idea but I think you shouldn't put it in the `__mocks__` folder. you can create a `./src/helpers/defaults` file and include things like this.

File: src/validation/emailValidation.js [hunk 26]
```diff
@@ -0,0 +1,9 @@
...
+
+export const emailRegex = (email) => {
+  // eslint-disable-next-line no-useless-escape
```
Comment (rovilay) [inline src/validation/emailValidation.js:5]: why this disable `no-useless-escape`

File: src/__tests__/emailverification.js (hunk 24 above)
Comment (zanio) [reply src/__tests__/emailverification.js]: 400. That was an oversight

File: src/__tests__/emailverification.js (hunk 23 above)
Comment (zanio) [reply src/__tests__/emailverification.js:66]: Thanks

File: .gitignore (hunk 20 above)
Comment (zanio) [reply .gitignore]: Fixed

File: package.json (hunk 21 above)
Comment (zanio) [reply package.json:7]: `-r dotenv/config ` it makes the `dotenv.config()` environment available in our test environment

File: src/__tests__/emailverification.js (hunk 22 above)
Comment (zanio) [reply src/__tests__/emailverification.js]: Done

File: src/validation/emailValidation.js (hunk 26 above)
Comment (zanio) [reply src/validation/emailValidation.js:5]: It is not a useless escape as it would affect my regex if i remove the escape character. The escape character makes it possible to be able to write more regex.

File: src/validation/emailValidation.js (hunk 25 above)
Comment (zanio) [reply src/validation/emailValidation.js]: Alright, it has been done!

File: src/index.js (hunk 16 above)
Comment (rovilay) [reply src/index.js:22]: Ok good.

File: src/middlewares/mail.js (hunk 18 above)
Comment (rovilay) [reply src/middlewares/mail.js:51]: @zanio 👆

File: src/routes/api/users.js [hunk 27]
```diff
@@ -1,82 +1,12 @@
...
-    .then((user) => {
-      if (!user) {
-        return res.sendStatus(401);
-      }
-      return res.json({ user: user.toAuthJSON() });
-    })
-    .catch(next);
-});
+router.post('/users/email/test', handleEmptyEmailBody, handleInvalidEmail,
+  SendVerificationEmail, emailverification.signUp);
```
[the remark this answers is not in the document]
Comment (rovilay) [reply src/routes/api/users.js:9]: Please add new line before starting a new function.

File: src/routes/api/users.js [hunk 28]
```diff
@@ -1,82 +1,12 @@
...
-  User.findById(req.payload.id)
-    .then((user) => {
-      if (!user) {
-        return res.sendStatus(401);
-      }
-      return res.json({ user: user.toAuthJSON() });
-    })
-    .catch(next);
-});
+router.post('/users/email/test', handleEmptyEmailBody, handleInvalidEmail,
```
[the remark this answers is not in the document]
Comment (rovilay) [reply src/routes/api/users.js:7]: Please add new line before starting a new function.

File: .env.example (hunk 19 above)
Comment (rovilay) [reply .env.example]: Can you please include the description too?
reach out to @conquext

File: package.json (hunk 21 above)
Comment (rovilay) [reply package.json:7]: ok good.

File: .env.example (hunk 19 above)
Comment (rovilay) [reply .env.example]: Please include the description next to the `env variable`

File: src/helper/default.js [hunk 29]
```diff
@@ -0,0 +1,3 @@
+/* eslint-disable import/prefer-default-export */
```
Comment (rovilay) [inline src/helper/default.js]: I specifically said `helpers` not `helper`. Please update

**Comment by rovilay** [comment]
@zanio I tried the `/event/test/` api and I recieve a success response on postman but I do not get a mail.

File: src/helper/default.js (hunk 29 above)
Comment (zanio) [reply src/helper/default.js]: Okay . done

File: .env.example (hunk 19 above)
Comment (zanio) [reply .env.example]: Done

File: src/routes/api/users.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (zanio) [reply src/routes/api/users.js:7]: Done

**Comment by zanio** [comment]
> @zanio I tried the `/event/test/` api and I recieve a success response on postman but I do not get a mail.

@rovilay It has been resolved. Please try again. it because we are using gmail we need to enable the `Less secure app access` check your spam folder as well

image

File: src/routes/api/users.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (zanio) [reply src/routes/api/users.js:9]: done

File: src/routes/api/users.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (zanio) [reply src/routes/api/users.js:7]: Done

File: src/routes/api/users.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (zanio) [reply src/routes/api/users.js:9]: Done

File: src/middlewares/mail.js (hunk 18 above)
Comment (zanio) [reply src/middlewares/mail.js:51]: The code has been updated. it is now 403.

**Comment by rovilay** [comment]
I get this when I click on verify `http://localhost/api/v1/users/email/verify?id=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InNvbWVfZW5jb2RlZF9pZGVudGlpdHkiLCJpYXQiOjE1NjcxMDA4MTYsImV4cCI6MTU2NzE4NzIxNn0.rL8q57NiAPxJ4O901Oom0k-khPHk-bGlag4XXgB2wuA`
No port

**Comment by rovilay** [comment]
your tests are failing please check.
@rovilay it has been fixed as well

**Comment by zanio** [comment]
> I get this when I click on verify `http://localhost/api/v1/users/email/verify?id=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InNvbWVfZW5jb2RlZF9pZGVudGlpdHkiLCJpYXQiOjE1NjcxMDA4MTYsImV4cCI6MTU2NzE4NzIxNn0.rL8q57NiAPxJ4O901Oom0k-khPHk-bGlag4XXgB2wuA`
> No port

@rovilay it has been fixed

File: src/controllers/emailController.js [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+ * @returns {object} Success email response for email token sent to mail
+ */
+  static confirmEmailVerificaionToken(req, res) {
+    const { id } = url.parse(req.url, true).query;
+
+    try {
+      jwt.verify(id, process.env.SECRET_KEY_EMAIL_VERIFY_TOKEN);
+      return res.status(200).json({
+        status: 200,
+        data: { message: 'You Account has been successfully verified, you would be redirected in few seconds to your dashboard' }
```
Comment (rovilay) [inline src/controllers/emailController.js]: Please remove the `you would be redirected in few seconds to your dashboard`  since the user isn't actually being redirected to the dashboard.

Comment (rovilay) [inline src/controllers/emailController.js]: `your` not `You`.

Comment (zanio) [reply src/controllers/emailController.js]: Done

Comment (zanio) [reply src/controllers/emailController.js]: Done

**Review by rovilay — APPROVED** [review]
